File: src/core/find/Packet.ts
import { Device } from "./index.js";

export class Packet {
    private readonly DATA_LENGTH_BYTE = 2;
    private readonly payload: Buffer;
    private data: Buffer;
    private readonly type: string;
    private device = new Device();
    private readonly days = {
        "1": "Monday",
        "2": "Tuesday",
        "3": "Wednesday",
        "4": "Thursday",
        "5": "Friday",
        "6": "Saturday",
        "7": "Sunday",
    };

    constructor(payload: Buffer) {
        this.payload = payload;
        this.data = this.getData();
        this.type = this.data.toString("utf8", 1, 2);
    }

    private getData(): Buffer {
        const { payload } = this;

        const length = payload[this.DATA_LENGTH_BYTE]!;
        const start = this.DATA_LENGTH_BYTE + 1;
        const end = start + length;

        const data = payload.subarray(start, end);

        if (data.length !== length) {
            throw new Error("Invalid data length");
        }

        return data;
    }

    public getType(): string {
        return this.type;
    }

    public parse(): Packet {
        this.device.system = this.getSystem();
        this.device.date.numeric = this.getDateNumeric();
        this.device.time = this.getTime();
        this.device.day = this.days[this.getDay() as keyof typeof this.days];
        this.device.date.text = this.getDateText();
        this.device.mac = this.getMac();
        this.device.hostname = this.getHostname();
        this.device.id = this.getId();

        return this;
    }

    public getDevice(): Device {
        return this.device;
    }

    private getDay(): string {
        const day = this.data[0]!.toString();

        this.data = this.data.subarray(3);

        return day;
    }

    private getDateNumeric(): string {
        const month = this.data[0]!.toString().padStart(2, "0");
        const day = this.data[1]!.toString().padStart(2, "0");
        const year = this.data.subarray(2, 4).readInt16BE();

        this.data = this.data.subarray(4);

        return `${month}/${day}/${year}`;
    }

    private getEnd(start: number = 0, end: string = "\0"): number {
        const { data } = this;

        const length = data.indexOf(end, start) - start;

        return start + length;
    }

    private getDateText(): string {
        const end = this.getEnd(0, "\0\0\0");
        const date = this.data.subarray(0, end).toString();

        this.data = this.data.subarray(end + 3);

        return date;
    }

    private getSystem(): number {
        const system = this.data.subarray(2, 4).readInt16BE();
        this.data = this.data.subarray(21);

        return system;
    }

    private getHostname(): string {
        const end = this.getEnd(0);
        const hostname = this.data.subarray(0, end).toString();

        this.data = this.data.subarray(end + 1);

        return hostname;
    }

    private getId(): string {
        const end = this.getEnd(0);

        return this.data.subarray(0, end).toString();
    }

    private getTime(): string {
        const hours = this.data[0]!.toString().padStart(2, "0");
        const minutes = this.data[1]!.toString().padStart(2, "0");
        const seconds = this.data[3]!.toString().padStart(2, "0");

        this.data = this.data.subarray(3);

        return `${hours}:${minutes}:${seconds}`;
    }

    private getMac(): string {
        const mac: Array<string> = [];

        for (let i = 0; i < 6; i++) {
            mac.push(this.data[i]!.toString(16).padStart(2, "0"));
        }

        this.data = this.data.subarray(6);

        return mac.join(":");
    }
}

File: go/internal/find/packet.go
package find

import (
	"bytes"
	"encoding/binary"
	"fmt"
)

var days = [8]string{
	0: "",
	1: "Monday",
	2: "Tuesday",
	3: "Wednesday",
	4: "Thursday",
	5: "Friday",
	6: "Saturday",
	7: "Sunday",
}

// minFixedLen is the number of bytes consumed by all fixed-length fields
// (system 21 + date-numeric 4 + time/day 7 + MAC 6).
const minFixedLen = 38

type Packet struct {
	data   []byte
	cursor int
}

func NewPacket(payload []byte) (*Packet, error) {
	if len(payload) < 3 {
		return nil, fmt.Errorf("payload too short")
	}

	length := int(payload[2])
	start := 3
	end := start + length

	if len(payload) < end {
		return nil, fmt.Errorf("data length exceeds payload")
	}

	return &Packet{
		data: payload[start:end],
	}, nil
}

func (p *Packet) at(offset int) byte {
	return p.data[p.cursor+offset]
}

func (p *Packet) int16BE(offset int) int16 {
	return int16(binary.BigEndian.Uint16(p.data[p.cursor+offset : p.cursor+offset+2]))
}

func (p *Packet) advance(n int) {
	p.cursor += n
}

func (p *Packet) remaining() []byte {
	return p.data[p.cursor:]
}

func (p *Packet) Parse() (*Device, error) {
	if len(p.data) < minFixedLen {
		return nil, fmt.Errorf("packet too short (have %d, need at least %d bytes)", len(p.data), minFixedLen)
	}

	d := &Device{}

	// getSystem(): reads data[2..4] as Int16BE, advances 21.
	d.System = int(p.int16BE(2))
	p.advance(21)

	// getDateNumeric(): month[0], day[1], year[2..4], advances 4.
	d.Date.Numeric = fmt.Sprintf("%02d/%02d/%d", p.at(0), p.at(1), p.int16BE(2))
	p.advance(4)

	// getTime(): hours[0], minutes[1], seconds[3], advances 3.
	// Note: data[3] is read for seconds but the cursor only moves 3 bytes,
	// leaving data[3] as data[0] for the next call (getDay).
	d.Time = fmt.Sprintf("%02d:%02d:%02d", p.at(0), p.at(1), p.at(3))
	p.advance(3)

	// getDay(): reads data[0] (= old data[3] from time block), advances 3.
	dayVal := p.at(0)
	if int(dayVal) < len(days) && days[dayVal] != "" {
		d.Day = days[dayVal]
	} else {
		d.Day = fmt.Sprintf("%d", dayVal)
	}

	p.advance(3)

	// getDateText(): read until triple-null '\0\0\0', advance past content + 3.
	endIdx := bytes.Index(p.remaining(), []byte{0, 0, 0})
	if endIdx < 0 {
		return nil, fmt.Errorf("packet missing date text terminator")
	}

	d.Date.Text = string(p.remaining()[:endIdx])
	p.advance(endIdx + 3)

	// getMac(): 6 bytes as colon-separated lower-case hex, advances 6.
	if p.cursor+6 > len(p.data) {
		return nil, fmt.Errorf("packet too short for MAC")
	}

	d.MAC = fmt.Sprintf("%02x:%02x:%02x:%02x:%02x:%02x",
		p.at(0), p.at(1), p.at(2), p.at(3), p.at(4), p.at(5))
	p.advance(6)

	// getHostname(): read until '\0', advance past content + 1.
	nullIdx := bytes.IndexByte(p.remaining(), 0)
	if nullIdx < 0 {
		return nil, fmt.Errorf("packet missing hostname terminator")
	}

	d.Hostname = string(p.remaining()[:nullIdx])
	p.advance(nullIdx + 1)

	// getId(): read until '\0' (or end of buffer).
	nullIdx = bytes.IndexByte(p.remaining(), 0)
	if nullIdx < 0 {
		d.ID = string(p.remaining())
	} else {
		d.ID = string(p.remaining()[:nullIdx])
	}

	return d, nil
}
